stl/inc/xstring

```diff
@@ -2546,79 +2660,206 @@ public:
+                _Traits::move(_My_data._Bx._Buf, _Arg, _Count);
+                _Traits::assign(_My_data._Bx._Buf[_Count], _Elem());
+            } else { // _Strat == _Construct_strategy::_From_string
+#ifdef _INSERT_STRING_ANNOTATION
+                _Traits::move(_My_data._Bx._Buf, _Arg, _Count);
+#else // ^^^ _INSERT_STRING_ANNOTATION ^^^ // vvv !_INSERT_STRING_ANNOTATION vvv
+                _Traits::move(_My_data._Bx._Buf, _Arg, _BUF_SIZE);
+#endif // !_INSERT_STRING_ANNOTATION
+            }
+
+            _ASAN_STRING_CREATE(*this);
+            _Proxy._Release();
+            return;
+        }
+
+        const size_type _New_capacity = _Calculate_growth(_Count);
+        const pointer _New_ptr        = _Al.allocate(_New_capacity + 1); // throws
+        _Construct_in_place(_My_data._Bx._Ptr, _New_ptr);
+
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) { // Begin the lifetimes of the objects before copying to avoid UB
+            _Traits::assign(_Unfancy(_New_ptr), _New_capacity + 1, _Elem());
+        }
+#endif // _HAS_CXX20
+
+        _My_data._Mysize = _Count;
+        _My_data._Myres  = _New_capacity;
+        if constexpr (_Strat == _Construct_strategy::_From_char) {
+            _Traits::assign(_Unfancy(_New_ptr), _Count, _Arg);
+            _Traits::assign(_Unfancy(_New_ptr)[_Count], _Elem());
+        } else if constexpr (_Strat == _Construct_strategy::_From_ptr) {
+            _Traits::copy(_Unfancy(_New_ptr), _Arg, _Count);
+            _Traits::assign(_Unfancy(_New_ptr)[_Count], _Elem());
+        } else { // _Strat == _Construct_strategy::_From_string
+            _Traits::copy(_Unfancy(_New_ptr), _Arg, _Count + 1);
+        }
+
+        _ASAN_STRING_CREATE(*this);
+        _Proxy._Release();
     }
 
     template <class _Iter>
-    _CONSTEXPR20 void _Construct(const _Iter _First, const _Iter _Last, forward_iterator_tag) {
-        // initialize from [_First, _Last), forward iterators
-        const size_type _Count = _Convert_size<size_type>(static_cast<size_t>(_STD distance(_First, _Last)));
-        reserve(_Count);
-        _Construct(_First, _Last, input_iterator_tag{});
-    }
+    _CONSTEXPR20 void _Construct_from_iter(_Iter _First, const _Iter _Last) {
+        auto& _My_data  = _Mypair._Myval2;
+        auto& _Al       = _Getal();
+        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Al);
+        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _My_data);
 
-    _CONSTEXPR20 void _Construct(_Elem* const _First, _Elem* const _Last, random_access_iterator_tag) {
-        // initialize from [_First, _Last), pointers
-        if (_First != _Last) {
-            assign(_First, _Convert_size<size_type>(static_cast<size_t>(_Last - _First)));
+        _My_data._Mysize = 0;
+        _My_data._Myres  = _BUF_SIZE - 1;
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            _Construct_in_place(_My_data._Bx);
+            _My_data._Myres = _BUF_SIZE; // TRANSITION: constexpr SSO

```

Comment:
Currently we are in SSO when `_Myres == _BUF_SIZE` So this always choses at least the smallest size that gives use a large string.

Note that we will later calculate the real size if the string is larger but we will never fall back to small string if it would fit

Reply:
Ah, I see